### PR TITLE
Enable gem to authenticate remotely

### PR DESCRIPTION
**Pull request by awead**
Incorporates the Devise-related code from Scholarsphere into a Rails engine that interacts with a remote authentication service.

Since much of the gem's functionality relies on Devise, I've modeled it a lot based on it, even including the name itself.

### CODE REVIEW

File: app/controllers/devise_remote/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+module DeviseRemote
+  class SessionsController < ApplicationController
```
Comment (carolyncole) [inline app/controllers/devise_remote/sessions_controller.rb:4]: Should this be a module so you can easily override the before_destroy?

File: lib/devise_remote/http_header_authenticatable_behavior.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    # @return [Boolean]
+    # @note Called if the user doesn't already have a rails session cookie
+    def valid_user?(headers)
+      remote_user(headers).present?
+    end
+
+    protected
+
+      # @note In production, only check for REMOTE_USER. HTTP_ is removed from the variable before
+      #   it is passed to the application. In test or development, this may or may not
```
Comment (carolyncole) [inline lib/devise_remote/http_header_authenticatable_behavior.rb:30]: You may want to add something like 'This allows for modheades and other browser plugins to be used in development mode'

File: lib/generators/devise_remote/templates/add_ldap_attributes_template.erb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+class AddLdapAttributesTo<%= table_name.camelize %> < ActiveRecord::Migration<%= migration_version %>
+  def self.up
+    add_column :<%= table_name %>, :display_name, :string
```
Comment (carolyncole) [inline lib/generators/devise_remote/templates/add_ldap_attributes_template.erb]: Are these fields needed/wanted for this gem?  Should groups be added if we do want them?

**Review by carolyncole — COMMENTED** [review]
Looks good.  Just a couple comments.

File: app/controllers/devise_remote/sessions_controller.rb (hunk 1 above)
Comment (awead) [reply app/controllers/devise_remote/sessions_controller.rb:4]: @cam156 I thought about that, but decided the `prepend` approach might be better. What do you think? Either way, I should add some documentation.

File: lib/generators/devise_remote/templates/add_ldap_attributes_template.erb (hunk 3 above)
Comment (awead) [reply lib/generators/devise_remote/templates/add_ldap_attributes_template.erb]: @cam156 yeah, I'm not so sure either. I'm leaning towards taking them out...

File: app/controllers/devise_remote/sessions_controller.rb (hunk 1 above)
Comment (carolyncole) [reply app/controllers/devise_remote/sessions_controller.rb:4]: I keep forgetting about prepends!

Comment (awead) [reply app/controllers/devise_remote/sessions_controller.rb:4]: @cam156 I know! So do I!!! Old habits..

**Pull request merged**